xarray/core/combine.py

```diff
@@ -323,16 +324,17 @@ def _dataarray_concat(arrays, dim, data_vars, coords, compat,
         raise ValueError('data_vars is not a valid argument when '
                          'concatenating DataArray objects')
 
-    datasets = []
-    for n, arr in enumerate(arrays):
-        if n == 0:
-            name = arr.name
-        elif name != arr.name:
-            if compat == 'identical':
-                raise ValueError('array names not identical')
-            else:
-                arr = arr.rename(name)
-        datasets.append(arr._to_temp_dataset())
+    name = result_name(arrays)
+    if name is None and compat == 'identical':
+        raise ValueError('array names not identical')
```

Comment:
I think it’s also possible to get a name of None if none of the arguments had names at all. So this logic probably needs to go in result_name()

Reply:
None of the other `result_name` calls would use this logic, so I've implemented it but in `_dataarray_combine`.